File: a2a-with-agentcore/agents/biomechanics-lab/src/agent.ts
/**
 * Biomechanics Lab Agent - LangGraph Implementation
 *
 * An exercise physiology expert that provides structured workout routines
 * based on specific physiological goals.
 */

import { createReactAgent } from "@langchain/langgraph/prebuilt";
import type { CompiledStateGraph } from "@langchain/langgraph";
import { ChatBedrockConverse } from "@langchain/aws";
import { HumanMessage, AIMessage } from "@langchain/core/messages";
import { searchExercisesTool } from "./tools/searchExercises.js";

const SYSTEM_PROMPT = `You are an expert in exercise physiology and strength & conditioning.

Objective: Provide structured workout routines based on specific physiological goals (hypertrophy, endurance, power, strength).

Instructions:
1. When you receive a goal from the Orchestrator, analyze what type of workout is needed
2. Use the search_exercises tool to find appropriate exercises
3. If muscle group is specified, search for that specific group
4. If equipment constraints are mentioned, include them in the search
5. Return a structured JSON workout plan

If the Orchestrator asks for a "compromise" or adjustment:
- Prioritize intensity over duration to maintain the training stimulus
- Suggest exercise modifications that achieve similar muscle activation
- Reduce volume while maintaining exercise quality

Your response should always include a structured workout in this format:
{
  "workout": {
    "name": "Workout Name",
    "estimatedDuration": 45,
    "exercises": [
      {
        "id": "exercise-id",
        "name": "Exercise Name",
        "muscleGroup": "target-muscle",
        "equipment": ["required", "equipment"],
        "sets": 4,
        "reps": "8-12",
        "restSeconds": 90,
        "notes": "Form cues"
      }
    ]
  }
}

Tone: Clinical, data-driven, and technical.`;

// Create the LangGraph agent
export function createBiomechanicsAgent(): CompiledStateGraph<any, any, any> {
  const model = new ChatBedrockConverse({
    model: process.env.MODEL_ID || "us.amazon.nova-lite-v1:0",
    region: process.env.AWS_REGION || "us-east-1",
    maxTokens: 2000,
    temperature: 0.3,
  });

  const agent = createReactAgent({
    llm: model,
    tools: [searchExercisesTool],
  });

  return agent;
}

export interface WorkoutRequest {
  goal: string;
  constraints?: {
    duration?: number;
    equipment?: string[];
    difficulty?: "beginner" | "intermediate" | "advanced";
    muscleGroups?: string[];
  };
  isCompromise?: boolean;
}

export interface WorkoutResponse {
  workout: {
    name: string;
    estimatedDuration: number;
    exercises: Array<{
      id: string;
      name: string;
      muscleGroup: string;
      equipment: string[];
      sets: number;
      reps: string;
      restSeconds: number;
      notes?: string;
    }>;
  };
  reasoning?: string;
}

export async function generateWorkout(
  request: WorkoutRequest
): Promise<WorkoutResponse> {
  const agent = createBiomechanicsAgent();

  // Build the prompt based on the request
  let prompt = `Create a workout plan for the following goal: ${request.goal}`;

  if (request.constraints) {
    if (request.constraints.duration) {
      prompt += `\nTime available: ${request.constraints.duration} minutes`;
    }
    if (request.constraints.equipment && request.constraints.equipment.length > 0) {
      prompt += `\nAvailable equipment: ${request.constraints.equipment.join(", ")}`;
    }
    if (request.constraints.difficulty) {
      prompt += `\nDifficulty level: ${request.constraints.difficulty}`;
    }
    if (request.constraints.muscleGroups && request.constraints.muscleGroups.length > 0) {
      prompt += `\nTarget muscle groups: ${request.constraints.muscleGroups.join(", ")}`;
    }
  }

  if (request.isCompromise) {
    prompt += `\n\nIMPORTANT: This is a compromise request. The original workout couldn't be done due to constraints. Prioritize intensity over duration. Find alternatives that work within the given equipment and time constraints.`;
  }

  prompt += `\n\nUse the search_exercises tool to find appropriate exercises, then return a structured workout plan as JSON.`;

  // Run the agent
  const result = await agent.invoke({
    messages: [
      {
        role: "system",
        content: SYSTEM_PROMPT,
      },
      new HumanMessage(prompt),
    ],
  });

  // Extract the final response
  const lastMessage = result.messages[result.messages.length - 1];
  let responseText = "";

  if (lastMessage instanceof AIMessage) {
    if (typeof lastMessage.content === "string") {
      responseText = lastMessage.content;
    } else if (Array.isArray(lastMessage.content)) {
      responseText = lastMessage.content
        .filter((part): part is { type: "text"; text: string } =>
          typeof part === "object" && part.type === "text"
        )
        .map((part) => part.text)
        .join("");
    }
  }

  // Parse the workout from the response
  const jsonMatch = responseText.match(/\{[\s\S]*"workout"[\s\S]*\}/);
  if (jsonMatch) {
    try {
      const parsed = JSON.parse(jsonMatch[0]);
      return parsed as WorkoutResponse;
    } catch {
      // If JSON parsing fails, create a default response
    }
  }

  // Fallback response if parsing fails
  return {
    workout: {
      name: "Custom Workout",
      estimatedDuration: request.constraints?.duration || 45,
      exercises: [],
    },
    reasoning: responseText,
  };
}

// For streaming responses
export async function* streamWorkout(
  request: WorkoutRequest
): AsyncGenerator<string> {
  const agent = createBiomechanicsAgent();

  let prompt = `Create a workout plan for: ${request.goal}`;
  if (request.constraints) {
    if (request.constraints.duration) {
      prompt += ` (${request.constraints.duration} min)`;
    }
    if (request.constraints.equipment) {
      prompt += ` using ${request.constraints.equipment.join(", ")}`;
    }
  }

  if (request.isCompromise) {
    prompt += ". This is a compromise request - prioritize intensity over duration.";
  }

  const stream = await agent.stream(
    {
      messages: [
        { role: "system", content: SYSTEM_PROMPT },
        new HumanMessage(prompt),
      ],
    },
    { streamMode: "messages" }
  );

  for await (const [message] of stream) {
    const m = message as AIMessage;
    if (m.getType?.() === "ai" && !(m as any).tool_call_chunks?.length) {
      if (typeof m.content === "string") {
        yield m.content;
      } else if (Array.isArray(m.content)) {
        for (const part of m.content) {
          if (typeof part === "object" && part.type === "text") {
            yield part.text;
          }
        }
      }
    }
  }
}

File: a2a-with-agentcore/agents/life-sync/src/agent.py
"""Life Sync Agent - LangGraph Implementation.

A pragmatic logistics and lifestyle coordinator that validates
workout plans against real-world constraints.
"""

import json
from dataclasses import dataclass
from typing import Any, AsyncGenerator

from langchain_aws import ChatBedrockConverse
from langchain_core.messages import HumanMessage, AIMessage, SystemMessage
from langgraph.prebuilt import create_react_agent

from tools import get_calendar_availability, get_equipment_inventory, check_equipment_for_workout

SYSTEM_PROMPT = """You are a pragmatic logistics and lifestyle coordinator.

Objective: Validate plans against the user's real-world constraints:
time, equipment, and fatigue.

Instructions:
1. When the Orchestrator sends a proposed workout plan, check it against the user's schedule
2. Use get_calendar_availability to check if the user has enough free time
3. Use get_equipment_inventory or check_equipment_for_workout to verify equipment availability
4. Explicitly flag any conflicts found

Types of conflicts to check for:
- TIME: "user only has a 30-minute gap" or "no free time today"
- EQUIPMENT: "current location has no barbell" or "missing squat rack"
- Both conflicts may exist simultaneously

Your response should always include a structured analysis in this format:
{
  "analysis": {
    "hasConflicts": true/false,
    "conflicts": [
      {
        "type": "time" or "equipment",
        "severity": "high", "medium", or "low",
        "message": "Description of the conflict",
        "suggestion": "How to resolve or work around it"
      }
    ],
    "recommendation": "Overall recommendation for the Orchestrator"
  }
}

Tone: Empathetic, realistic, and brief."""


def create_life_sync_agent():
    """Create the Life Sync LangGraph agent."""
    model = ChatBedrockConverse(
        model=os.environ.get("MODEL_ID", "us.amazon.nova-lite-v1:0"),
        region_name=os.environ.get("AWS_REGION", "us-east-1"),
        max_tokens=2000,
        temperature=0.3,
    )

    tools = [
        get_calendar_availability,
        get_equipment_inventory,
        check_equipment_for_workout,
    ]

    agent = create_react_agent(
        model,
        tools,
    )

    return agent


import os


@dataclass
class WorkoutPlan:
    """A workout plan to validate."""
    name: str
    estimated_duration: int
    exercises: list[dict[str, Any]]

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "WorkoutPlan":
        return cls(
            name=data.get("name", "Unnamed Workout"),
            estimated_duration=data.get("estimatedDuration", 60),
            exercises=data.get("exercises", []),
        )


@dataclass
class ConflictAnalysis:
    """Analysis of conflicts for a workout plan."""
    has_conflicts: bool
    conflicts: list[dict[str, Any]]
    recommendation: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "analysis": {
                "hasConflicts": self.has_conflicts,
                "conflicts": self.conflicts,
                "recommendation": self.recommendation,
            }
        }


async def validate_workout(
    workout_plan: WorkoutPlan | None = None,
    raw_request: str | None = None,
    date: str | None = None,
    location: str | None = None,
) -> ConflictAnalysis:
    """Validate a workout plan against user constraints.

    Args:
        workout_plan: The structured workout plan to validate.
        raw_request: Raw text request if no structured plan.
        date: Date to check availability (YYYY-MM-DD).
        location: Location to check equipment.

    Returns:
        ConflictAnalysis with any detected conflicts.
    """
    agent = create_life_sync_agent()

    # Build the validation prompt
    if workout_plan:
        required_equipment = set()
        for exercise in workout_plan.exercises:
            for eq in exercise.get("equipment", []):
                required_equipment.add(eq)

        prompt = f"""Please validate this workout plan:

Workout: {workout_plan.name}
Duration: {workout_plan.estimated_duration} minutes
Exercises: {len(workout_plan.exercises)} exercises
Required Equipment: {', '.join(required_equipment) if required_equipment else 'None (bodyweight)'}

1. Check if the user has {workout_plan.estimated_duration} minutes available today
2. Check if all required equipment is available at their location ({location or 'home'})
3. Report any conflicts found

Return a structured analysis with hasConflicts, conflicts array, and recommendation."""
    else:
        prompt = raw_request or "Check my availability for a 60-minute workout today."

    # Run the agent
    result = await agent.ainvoke({
        "messages": [
            SystemMessage(content=SYSTEM_PROMPT),
            HumanMessage(content=prompt),
        ],
    })

    # Extract the response
    last_message = result["messages"][-1]
    response_text = ""
    if isinstance(last_message, AIMessage):
        if isinstance(last_message.content, str):
            response_text = last_message.content
        elif isinstance(last_message.content, list):
            response_text = " ".join(
                part.get("text", "") for part in last_message.content
                if isinstance(part, dict) and part.get("type") == "text"
            )

    # Parse the analysis from the response
    try:
        json_match = None
        # Try to find JSON in the response
        import re
        json_pattern = r'\{[\s\S]*"analysis"[\s\S]*\}'
        match = re.search(json_pattern, response_text)
        if match:
            parsed = json.loads(match.group())
            analysis = parsed.get("analysis", {})
            return ConflictAnalysis(
                has_conflicts=analysis.get("hasConflicts", False),
                conflicts=analysis.get("conflicts", []),
                recommendation=analysis.get("recommendation", response_text),
            )
    except json.JSONDecodeError:
        pass

    # Fallback: Parse from text
    has_conflicts = any(
        word in response_text.lower()
        for word in ["conflict", "missing", "unavailable", "no time", "limited"]
    )

    return ConflictAnalysis(
        has_conflicts=has_conflicts,
        conflicts=[],
        recommendation=response_text,
    )


async def stream_validation(
    workout_plan: WorkoutPlan | None = None,
    raw_request: str | None = None,
) -> AsyncGenerator[str, None]:
    """Stream the validation process.

    Yields chunks of the validation response.
    """
    agent = create_life_sync_agent()

    if workout_plan:
        prompt = f"Validate this workout: {workout_plan.name} ({workout_plan.estimated_duration} min)"
    else:
        prompt = raw_request or "Check my availability for a workout."

    async for event in agent.astream_events(
        {
            "messages": [
                SystemMessage(content=SYSTEM_PROMPT),
                HumanMessage(content=prompt),
            ],
        },
        version="v2",
    ):
        if event["event"] == "on_chat_model_stream":
            chunk = event["data"].get("chunk")
            if chunk and hasattr(chunk, "content"):
                content = chunk.content
                if isinstance(content, str):
                    yield content
                elif isinstance(content, list):
                    for part in content:
                        if isinstance(part, dict) and part.get("type") == "text":
                            yield part.get("text", "")
